### DISCUSSION TITLE
Serving SqlAlchemy MetaData Objects

### DISCUSSION THREAD

**Discussion by AndrewHannigan** [sqlalchemy/sqlalchemy#7112] [Ideas]
Suppose I have a data engineering team relying heavily sqlalchemy to maintain and interact with some databases.  What is the recommended way to manage a common SqlAlchemy metadata object representing those databases?  I can think of two options currently:

### Option 1 - Hardcode the Metadata object in your python package.  
The issue here is that if the schema changes in the future, your code may break because your metadata object is out of date.  Additionally, each user needs to maintain their own version of the MetaData object in their code, so you have multiple people spending time doing the same thing.

### Option 2 - Reflect the MetaData object at runtime.  
This avoids the need to maintain your own MetaData object in your package.  However, it takes up time at runtime that might take significantly longer than the task you actually wish to accomplish.  Also reflection may not always be perfect, and occasionally it may be important to override the inferred schema for correctness.  This can be done, but then we inherit some of the hardcoding problems of Option 1.

Both of these approaches have draw backs, and both suffer from a lack of single source of truth.  What would be great is a way to serve metadata objects from a centralized service, so users can focus on interacting with and manipulating SA table objects instead of maintaining them.  I could even imagine an integration with GCP's data catalog or AWS glue that builds SA metadata objects from those, so there is truly a single institutional source of truth.  What is the recommended approach to this problem currently?

**Comment by zzzeek** [comment]
I would choose 3. offer a single package as a dependency throughout the organization that packages this metadata, along with alembic migrations matching the databases in question.

**Reply by AndrewHannigan** [reply]
That would probably be an improvement on hardcoding within each python package.  But I think there is still an issue of staleness there just because it is ultimately python code that's being distributed.  For instance if I build a docker image that runs daily and does some manipulations/interactions with sqlalchemy, the image gets printed with whatever the metadata package is at that point in time that I built the docker image.  I'd have to rebuild the image every time there is a change to that metadata package.

**Reply by zzzeek** [reply]
i dont see how that is avoidable anyway, because the application code which calls upon this metadata we would assume refers to specific tables and columns.  if you use reflection, besides the slow startup time, if the code is looking for table/column "foo.bar" and that gets removed, you have to upgrade anyway.

I think this kind of thing is something to be decided on a case by case basis.  for short term ETL scripts, I would totally use reflection.   for comprehensive applications that are intricately architected around the shape of the schema, they have to be upgraded along with the database.

if you truly have a core set of SQL databases that are widely used throughout an organization, you actually don't want to have distant teams building MetaData around that at all. you want to provide a RESTful API with versioning.   the RESTful API is then where you have your SQLAlchemy code, close to the database.  distant services that have weak connections to these databases can instead build to RESTful APIs that thanks to versioning will change much more slowly than the schema itself.

**Reply by CaselIT** [reply]
Personally I think any such solution would be out of scope of sqlalchemy the repository, but may be implemented by a 3rd party package.

>  if I build a docker image that runs daily and does some manipulations/interactions with sqlalchemy, the image gets printed with whatever the metadata package is at that point in time that I built the docker image. I'd have to rebuild the image every time there is a change to that metadata package.

I would think that yes, each time the package releases a version a new image is created.

Otherwise a solution could be to just add a `pip install -U your_metadata_package_here -i your/index/url` before each run, that would be equivalent to the hypothetical 3rd party service that gets queried on script start

**Reply by AndrewHannigan** [reply]
Thanks for the insight. I do agree that if it's something worth doing it would belong in a 3rd party package and not necessarily in SA itself. I may also need to think a bit more deeply about the actual benefit of something like this given the points raised. Will report back if anything like a new package comes of it.
